### PR TITLE
WIP: Add experimental command to install/remove OpenShift on Kube

### PR DESCRIPTION
**Pull request by derekwaynecarr**
This command is functional, but could use some clean-up.

To use the command, you need the following:
- a kubernetes cluster
- the public key file for that cluster to secure service accounts
- the cluster must support LoadBalancer services

So to use against Kubernetes on GCE:

```
# spin up a kubernetes cluster
$ cd go/src/k8s.io/kubenetes
$ make quick-release
$ export KUBERNETES_PROVIDER=gce
$ export KUBE_GCE_INSTANCE_PREFIX="derek"
$ cluster/kube-up.sh

# get the key
$ export ZONE=$(gcloud compute instances list | grep "${KUBE_GCE_INSTANCE_PREFIX}\-master" | awk '{print $2}' | head -1)
$ echo "sudo cat /srv/kubernetes/server.key; exit;" | gcloud compute ssh ${KUBE_GCE_INSTANCE_PREFIX}-master --zone ${ZONE} | grep -Ex "(^\-.*\-$|^\S+$)" > /tmp/serviceaccounts.private.key

# deploy openshift into the demo namespace
$ openshift ex on-kubernetes install \
--kubeconfig=/home/decarr/.kube/config \
--namespace=demo \
--service-account-public-key=/tmp/serviceaccounts.private.key

# clean-up openshift
$ openshift ex on-kubernetes remove \
--namespace=demo \
--kubeconfig=/home/decarr/.kube/config
```

Things this command could use:
- actual output to the terminal explaining what happens!
- ability to specify ephemeral or pvc backed storage for etcd
- populate default templates

/cc @smarterclayton @csrwng @bparees @paralin

### CODE REVIEW

**Comment by paralin** [comment]
Looks good, as a reference use http://github.com/paralin/openshift-under-kubernetes

It now supports:

 - Full deploy of OpenShift including PV and PVCs with dynamically created volumes
 - Output explaining what's happening
 - Population of default templates and automatic configuration
 - Address various random problems I've encountered trying to do these deploys
 - Commands to edit config, get config locally, add a cluster role to a user
 - Run administrative commands inside a pod, for example, diagnostics, reconcile roles (used in my internal deploy scripts)
 - Remove from a cluster
 - Deploy an OpenShift registry without using OpenShift resources (pure Rc and service) with dynamically allocated PV and PVC

It's good to upstream this stuff. Still a lot to add though.

**Comment by derekwaynecarr** [comment]
@paralin 

>>Full deploy of OpenShift including PV and PVCs with dynamically created volumes

I am interested in this, definitely a must do before it goes out of experimental.

>>Output explaining what's happening

I definitely need to do this.

>>Population of default templates and automatic configuration

I had hoped this was going to be happening by default now as part of `openshift start`, but this may no longer be the case.  If not, I would like to do this because it improves out of the box experience.

>>Address various random problems I've encountered trying to do these deploys
Are you running etcd clusters or single node etcd?

>>Commands to edit config, get config locally, add a cluster role to a user
So write now, I clean-up the temp directory that holds the generated config, but this could be a flag.  In general, I want a `kubectl` command that can export a secret/configmap to an exploded set of files on my local disk.

>>Remove from a cluster
@liggitt @smarterclayton longer term, if we want to support nested deployments, we will need to have a way to tell OpenShift what it's finalizer token is in master-config.yaml

>>Deploy an OpenShift registry without using OpenShift resources (pure Rc and service) with dynamically allocated PV and PVC
Interesting.  I struggled even with the ReplicationController versus ReplicaSet versus Deployment options in general.  Moving this out of experimental will require versioned output.

@paralin - are you interested in collaborating on the command further?

**Comment by smarterclayton** [comment]
Population by default is planned.

**Comment by paralin** [comment]
@derekwaynecarr Sure. Our company (http://beta.faceit.com) is working towards this as a deployment model for dev and prod anyway, which is why I've been making tons of issues around it in the last few days.

**Comment by derekwaynecarr** [comment]
@smarterclayton @csrwng - I was running e2es, and had a moment to add some output.  Thoughts?

```
-- Check if namespace demo exists ... 
-- Created namespace
   OK
-- Check if service etcd-discovery exists ... 
-- Created service
   OK
-- Check if replication controller etcd-discovery exists ... 
-- Created replication controller
   OK
-- Check if service etcd exists ... 
-- Created service
   OK
-- Check if replication controller etcd exists ... 
-- Created replication controller
   OK
-- Check if service openshift exists ... 
-- Created service
   OK
-- Detecting public IP for service openshift ..................................... OK
-- Check if secret openshift exists ... 
-- Created secret
   OK
-- Check if replication controller openshift exists ... 
-- Created replication controller
   OK
-- Server Information ...
   OpenShift server starting...
   The server is accessible via web console at:
   https://104.154.27.11:8443
```

**Comment by smarterclayton** [comment]
Too much - we should sync but I think this is probably too wordy.

**Comment by csrwng** [comment]
looks familiar :)
I'd say you don't need to write out 2 different lines for the "checking blah" and then "creating blah" parts. Maybe "Ensure blah ... OK" would be better.

**Comment by marun** [comment]
@derekwaynecarr How will openshift be able to launch a pod when running under kube, or is that still to be determined?  I apologize if I'm missing something obvious.

**Comment by paralin** [comment]
@marun openshift calls the kubernetes api

**Comment by marun** [comment]
@paralin Are there docs you could point me to that explain the interaction?  My naive mental model is based on the following which doesn't seem to provide for api layering:

https://github.com/kubernetes/kubernetes/blob/master/docs/design/architecture.md

**Comment by paralin** [comment]
I don't know about docs but you can just add openshift anywhere in the
white space and draw a line to the kubernetes rest api.

**Comment by paralin** [comment]
Oh and when you use openshift under kube you have a choice of interacting
with the base kubernetes api directly or hitting the openshift api which
proxies anything the base kubernetes api can handle

**Comment by marun** [comment]
@paralin I'm afraid I'm still fuzzy on how responsibilities are divided.  What happens when openshift receives a request to create a pod?  Without layering, something like the following would happen:

 - api validates request and writes to etcd
 - controllers read etcd state and schedule to a kubelet
 - kubelet launches the pod

What's the sequence with openshift running on kube?  Do all requests to openshift just get proxied to kube?  Or given that this PR deploys etcd, does an api request to openshift update openshift's etcd and then openshift controllers attempt to update the kube cluster state?  It doesn't make sense to me how openshift and kube could maintain desired cluster state separately and coordinate its application.

**Comment by paralin** [comment]
@marun It receives a request to create a pod and it proxies the HTTP request to Kubernetes. End of line.

Etcd is used to store Openshift specific resources and controllers like Deployments, Builds, etc. These are just layers on top of existing resources like ReplicationControllers. So if you start a build, it will create a Build object in OpenShift's etcd store, call Kubernetes to spawn a pod to execute that build, and then monitor that pod state and propagate it back to the build state.

**Comment by derekwaynecarr** [comment]
Correct, OpenShift if started in this manner will proxy to Kubernetes when
dealing with Kubernetes specific resource types.  This means that those
objects are stored in the etcd instance supporting the underlying
Kubernetes cluster.  So namespaces, pods, services; etc all go there.
OpenShift has its own etcd running as a pod to persist it's additional
resources (builds, images, etc).  When interfacing with Kubernetes through
OpenShift, we proxy those calls to enforce our policy auth/authz pieces.

There are features in OpenShift that are not supported in this deployment
model.  Most notably,
SecurityContextConstraints are not supported which can be a big deal
depending on the security posture of a deployment.

**Comment by marun** [comment]
Thank you both for the explanation, makes sense now.

**Comment by paralin** [comment]
Where are we on this?

**Comment by tenfourty** [comment]
@derekwaynecarr are there plans to get this command fully working and for Origin to run on an external kubernetes cluster?

**Comment by tenfourty** [comment]
See also issue #10367 on this topic.
